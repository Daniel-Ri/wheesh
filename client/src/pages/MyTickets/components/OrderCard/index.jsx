import PropTypes from 'prop-types';
import { FormattedMessage } from 'react-intl';
import { useNavigate } from 'react-router-dom';

import { formatHour, formatOrderDate } from '@utils/handleValue';

import classes from './style.module.scss';

const OrderCard = ({ order }) => {
  const navigate = useNavigate();

  const handleClickCard = () => {
    if (!order.Payment.isPaid) {
      navigate(`/unpaid/${order.id}`);
    } else if (new Date(order.Schedule.arrivalTime) >= new Date(new Date() - 6 * 60 * 60 * 1000)) {
      navigate(`/order/${order.id}`);
    } else {
      navigate(`/history/${order.id}`);
    }
  };

  return (
    <div data-testid="OrderCard" className={classes.card} onClick={handleClickCard}>
      <div className={classes.stations}>
        <div className={classes.stationName}>{order.Schedule.departureStation.name}</div>
        <div className={classes.train}>
          <div className={classes.name}>{order.Schedule.Train.name}</div>
          <div className={classes.arrowImage}>
            <img src="/arrowTrain.png" alt="" />
          </div>
        </div>
        <div className={classes.stationName}>{order.Schedule.arrivalStation.name}</div>
      </div>
      <div className={classes.timeDesc}>
        <FormattedMessage id="app_departure" />: {formatHour(order.Schedule.departureTime)}{' '}
        {formatOrderDate(order.Schedule.departureTime)}
      </div>
      <div className={classes.timeDesc}>
        <FormattedMessage id="app_order_time" />: {formatHour(order.createdAt)} {formatOrderDate(order.createdAt)}
      </div>
      <div className={classes.tickets}>
        {order.OrderedSeats.length}{' '}
        {order.OrderedSeats.length > 1 ? <FormattedMessage id="app_tickets" /> : <FormattedMessage id="app_ticket" />}{' '}
        {order.OrderedSeats.map((orderedSeat) => orderedSeat.name).join(', ')}
      </div>
    </div>
  );
};

OrderCard.propTypes = {
  order: PropTypes.object.isRequired,
};

export default OrderCard;
