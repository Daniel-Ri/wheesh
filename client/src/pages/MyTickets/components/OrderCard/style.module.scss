@import '@styles/base';

.card {
  width: 100%;
  border-radius: 10px;
  padding: 0.5rem 0.75rem;
  background-color: $color-white;
  box-shadow: 0 0 3px 0 $color-light-gray;
  @include directionY;
  gap: 0.5rem;
  cursor: pointer;
  transition: 0.3s ease-in-out;

  .stations {
    @include directionX;
    align-items: center;
    gap: 1rem;

    .stationName {
      font-size: 90%;
    }

    .train {
      min-width: 3rem;
      @include directionY;
      align-items: center;

      .name {
        margin-top: -0.75rem;
        font-size: 70%;
      }

      .arrowImage {
        margin-top: -0.9rem;
        width: 100%;
        height: 0.5rem;

        img {
          width: 100%;
          height: 100%;
          object-fit: cover;
        }
      }
    }
  }
  
  .timeDesc {
    color: $color-cool-gray;
    font-size: 80%;
  }

  .tickets {
    color: $color-cool-gray;
    font-size: 80%;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
  }
}

.card:hover {
  box-shadow: 0 0 6px 0 $color-cool-gray;
}