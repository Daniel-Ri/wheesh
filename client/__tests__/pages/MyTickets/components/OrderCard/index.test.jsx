import { render } from '@utils/testHelper';
import OrderCard from '@pages/MyTickets/components/OrderCard';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

const mockProps = {
  order: {
    id: 11,
    userId: 1,
    scheduleId: 92,
    isNotified: false,
    createdAt: '2023-12-14T02:59:20.000Z',
    Payment: {
      id: 11,
      orderId: 11,
      amount: 900000,
      isPaid: false,
      duePayment: '2023-12-14T03:59:20.000Z',
    },
    Schedule: {
      id: 92,
      trainId: 2,
      departureStationId: 1,
      arrivalStationId: 3,
      departureTime: '2023-12-14T23:40:00.000Z',
      arrivalTime: '2023-12-15T00:10:00.000Z',
      departureStation: {
        id: 1,
        name: 'Halim',
      },
      arrivalStation: {
        id: 3,
        name: 'Padalarang',
      },
      Train: {
        id: 2,
        name: 'G1202',
      },
    },
    OrderedSeats: [
      {
        name: 'John Doe',
      },
      {
        name: 'Jean Doe',
      },
    ],
  },
};

beforeEach(() => {
  wrapper = render(<OrderCard {...mockProps} />);
});

describe('Component OrderCard of My Tickets page', () => {
  test('Renders OrderCard component', () => {
    const { getByTestId } = wrapper;
    expect(getByTestId('OrderCard')).toBeInTheDocument();
  });

  test('Should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
