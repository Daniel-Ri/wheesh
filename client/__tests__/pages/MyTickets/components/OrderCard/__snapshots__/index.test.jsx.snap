// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component OrderCard of My Tickets page Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="card"
        data-testid="OrderCard"
      >
        <div
          class="stations"
        >
          <div
            class="stationName"
          >
            Halim
          </div>
          <div
            class="train"
          >
            <div
              class="name"
            >
              G1202
            </div>
            <div
              class="arrowImage"
            >
              <img
                alt=""
                src="/arrowTrain.png"
              />
            </div>
          </div>
          <div
            class="stationName"
          >
            Padalarang
          </div>
        </div>
        <div
          class="timeDesc"
        >
          Keberangkatan
          : 
          06:40
           
          15/12/2023
        </div>
        <div
          class="timeDesc"
        >
          Waktu Pesanan
          : 
          09:59
           
          14/12/2023
        </div>
        <div
          class="tickets"
        >
          2
           
          tiket
           
          John Doe, Jean Doe
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="card"
      data-testid="OrderCard"
    >
      <div
        class="stations"
      >
        <div
          class="stationName"
        >
          Halim
        </div>
        <div
          class="train"
        >
          <div
            class="name"
          >
            G1202
          </div>
          <div
            class="arrowImage"
          >
            <img
              alt=""
              src="/arrowTrain.png"
            />
          </div>
        </div>
        <div
          class="stationName"
        >
          Padalarang
        </div>
      </div>
      <div
        class="timeDesc"
      >
        Keberangkatan
        : 
        06:40
         
        15/12/2023
      </div>
      <div
        class="timeDesc"
      >
        Waktu Pesanan
        : 
        09:59
         
        14/12/2023
      </div>
      <div
        class="tickets"
      >
        2
         
        tiket
         
        John Doe, Jean Doe
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
